test/vmtests/vmtests/utils/libvirt_utils.py

```diff
@@ -14,8 +18,99 @@ def __init__(self, name: str, memory_mib: int, core_count: int, os_disk_path: Pa
+        # Loop through all .json files in the folder
+        decoder = json.JSONDecoder()
+        firmware_configs = []
+        for firmware_definition_file in Path("/usr/share/qemu/firmware").glob("*.json"):
+            try:
+                with firmware_definition_file.open("r", encoding="utf-8") as f:
+                    data = f.read().lstrip()  # decode hates leading whitespace
+                    while data:
+                        obj, index = decoder.raw_decode(data)
+                        firmware_configs.append(obj)
+                        data = data[index:].lstrip()
+            except json.JSONDecodeError as e:
+                raise Exception(f"Error reading {firmware_definition_file.name}: {e}")
+
+        # Filter on architecture.
+        filtered_firmware_configs = list(
+            filter(lambda f: f["targets"][0]["architecture"] == arch, firmware_configs)
+        )
+
+        filtered_firmware_configs = list(
+            filter(
+                lambda f: any(
+                    fnmatch.fnmatch(full_machine_type, target_machine)
+                    for target_machine in f["targets"][0]["machines"]
+                ),
+                filtered_firmware_configs,
+            )
+        )
+
+        # Exclude Intel TDX and AMD SEV-ES firmwares.
+        filtered_firmware_configs = list(
+            filter(
+                lambda f: "inteltdx" not in f["mapping"]["executable"]["filename"]
+                and "amdsev" not in f["mapping"]["executable"]["filename"]
+                # qcow2 does azl2, need to exclude such entries
+                and "qcow2" not in f["mapping"]["executable"]["filename"],
+                filtered_firmware_configs,
+            )
+        )
+
+        # Filter on secure boot.
+        if enable_secure_boot:
+            filtered_firmware_configs = list(
+                filter(
+                    lambda f: "secure-boot" in f["features"]
+                    and "enrolled-keys" in f["features"],
+                    filtered_firmware_configs,
+                )
+            )
+        else:
+            filtered_firmware_configs = list(
+                filter(
+                    lambda f: "secure-boot" not in f["features"],
+                    filtered_firmware_configs,
+                )
+            )
+
+        # Get first matching firmware.
+        firmware_config = next(iter(filtered_firmware_configs), None)
+        if firmware_config is None:
+            raise LisaException(
+                f"Could not find matching firmware for machine-type={machine_type} "
+                f"({full_machine_type}) and secure-boot={enable_secure_boot}."
+            )
+
+        return firmware_config
+
 # Create XML definition for a VM.
-def create_libvirt_domain_xml(vm_spec: VmSpec) -> str:
+def create_libvirt_domain_xml(vm_spec: VmSpec, host_os: str, boot_type: str, uefi_firmware_binary: str) -> str:
```

Comment:
I would call `get_libvirt_firmware_config` in `create_libvirt_domain_xml` to avoid adding complexity to the individual tests.